ThirdPartyAdapters/chartboost/chartboost/src/main/java/com/google/ads/mediation/chartboost/ChartboostAdapter.java

```diff
@@ -186,6 +186,101 @@ public void onPause() {
     public void onResume() {
     }
 
+    @Override
+    public void requestBannerAd(Context context,
+                                MediationBannerListener mediationBannerListener,
+                                Bundle serverParameters,
+                                AdSize adSize,
+                                MediationAdRequest mediationAdRequest,
+                                Bundle networkExtras) {
+        mMediationBannerListener = mediationBannerListener;
+        mChartboostParams = ChartboostAdapterUtils.createChartboostParams(serverParameters, networkExtras);
+        if (!ChartboostAdapterUtils.isValidChartboostParams(mChartboostParams)) {
+            // Invalid server parameters, send ad failed to load event.
+            if (mMediationBannerListener != null) {
+                mMediationBannerListener.onAdFailedToLoad(ChartboostAdapter.this,
+                        AdRequest.ERROR_CODE_INVALID_REQUEST);
+            }
+            return;
+        }
+        ChartboostSingleton.startChartboostBanner(context, mChartboostBannerDelegate, createChartboostBannerListener());
```

Comment:
The value returned by `createChartboostBannerListener()` will be passed as the banner delegate in the end, which holds a weak reference to it right? Doesn't this mean that it will get deallocated and listener methods won't work?

Reply:
Each banner needs to have it's own listener. Due to little issue where banner needs to be created inside Singleton cause needs sdk to be initialized, and it requires listener which needs to be created inside adapter. When banner is detached or deallocated it will stop using listener.

Reply:
I still don't see how the listener is kept in memory. Does banner hold a strong reference to it?

Reply:
Listener is a strong reference. Solved